### ISSUE TITLE
Merge into master

### ISSUE THREAD

**Issue by fendor** [mpickering/haskell-ide-engine#48]
This issue tracks the progress to finally merge `hie-bios` into https://github.com/haskell/haskell-ide-engine.
It consists of all issues that need to be worked on or PRs that need to be merged before we are ready to merge. People may be assigned to the issues. These assignments are naturally non obligatory ;D. Also, work-stealing is promoted if you run out of tasks!

* [x] Update `hie-bios` ~after a release: https://github.com/mpickering/hie-bios/issues/88~ (@mpickering, supported by @fendor)
* [x] Update `haskell-lsp` after release to 0.18: https://github.com/alanz/haskell-lsp/pull/201 (@alanz and @bubba)
* [x] Update `haskell-lsp` to 0.19 (@fendor)
* [x] Update `cabal-helper` to latest (@bubba)
* [x] Merge #44 after `haskell-lsp-0.18` has been release (@mpickering)
* [x] Verify that #36 is indeed solved by #44 (hard to impossible. Currently can only be disproved, not proved) (everyone, we "want" to disprove it. If we can't, it might be fixed)
* [x] Merge #43 to fix HaRe standalone tests (@alanz and @mpickering )
* [x] Fix issue #34, the fix ought to be to catch the correct exception (@jneira, supported by @mpickering and @fendor)
* [x] Remove unused imports #53 
* [x] Dont show messages such as `lintCmd: no access to the persisted file.` https://github.com/mpickering/haskell-ide-engine/issues/36#issuecomment-551621937 (@fendor, closed by #51)
* [x] In the current master, multiple tests fail: https://circleci.com/workflow-run/544e0fd4-c7e8-4b2c-ab09-5e28159d2b1f 
    * [x] HaRe test (@alanz)
    * [x] Liquid haskell tests (solved by no particular commit, afaict)
    * [x] Other test-failures
* [x] Cabal-Helper: quality assurance and fixes
  * [x] Check that stack based projects also work with the implicit configuration  (@fendor and @jneira)
  * [x] Cabal-Helper: Select project context based on available tools #52 (@bubba)
  * [x]  Detect files in main-is of components for cabal-helper cradle #57 (@bubba)
  * [x]  Cabal-Helper: Ambiguous target 'test:tasty' #58  (@bubba)
  * ~[ ] Error in vscode: "`gcc.exe' failed in phase `C pre-processor'. (Exit code: 1)" #47~ not our problem, possible fix in https://github.com/haskell/cabal/pull/6441
  * ~[ ]  No Hover Responses on stack project #32~
* [x] Write Documentation for new functions (@mpickering and @fendor)
* [ ] Fix or remove all test-cases that are marked with `@fendor` (not crucial for merge)
     * [x] Definitions across components https://github.com/mpickering/haskell-ide-engine/blob/hie-bios/test/unit/HaRePluginSpec.hs#L189 (@alanz)
     * [ ] Behaviour on malformed projects https://github.com/mpickering/haskell-ide-engine/blob/hie-bios/test/functional/FunctionalBadProjectSpec.hs#L18
     * [ ] Clean up liquid haskell test-case. Maybe already covered in master. https://github.com/mpickering/haskell-ide-engine/blob/hie-bios/test/functional/FunctionalLiquidSpec.hs#L91 
* ~[ ] Case-split command has been removed: https://github.com/mpickering/haskell-ide-engine/blob/hie-bios/src/Haskell/Ide/Engine/Plugin/HaRe.hs#L315 reimplement it! (not crucial for merge)~ Will be done after merge.
* [x] Drop GHC 8.2.2 support. #44 by merging from upstream (@mpickering)
* [x] Integrate updates from upstream/master into hie-bios (last step to avoid that we have to repeat it several times) #82

This list is meant to be thourogh and complete. If something is missing, we will expand this list. It may look frightening but *almost* each of these issues is minor. We are almost done! 🐳

@bubba, @alanz, @mpickering, @jneira, @power-fungus, @lorenzo

**Comment by mpickering** [comment]
Can we assign people to each of these subtasks? Feel free to assign me to whatever you think I am best suited for.

**Comment by alanz** [comment]
I will take on the HaRe tests, and getting haskell-lsp prepared for release.

**Comment by lukel97** [comment]
> Check, if the user has the build-tools on the Path, before selecting a project type 

I can take a look at this @fendor

**Comment by lukel97** [comment]
An issue that's affecting my particular setup (I wouldn't let this block the release though) is that for cabal projects with with-compiler set to a different compiler than the `ghc` on the path, hie-bios fails as it just calls regular `ghc`. @dxld mentioned on IRC that c-h does return this information apparently so we could make use of this?

**Comment by fendor** [comment]
The ghc on the path is in my opinion an orthogonal problem to cradle selection. This can ultimately only be solved in `hie-bios` itself.
I will split it up into separate issues, there we can discuss what I meant with it.

**Comment by jneira** [comment]
I'm not sure if catching ghc exceptions would fix #32 cause i still dont know why hie doesnt respond within the `dhall-haskell` (and [it works in linux](https://github.com/mpickering/haskell-ide-engine/issues/32#issuecomment-540486748) but no in my windos systems). Anyway i think we shouldnt block merging if it is not fixed (i'll test it again with the last merged version just in case).
~~But it definitely could fix #47 and maybe #27.~~ (#47 is already showing the error in vscode, the problem is with `cabal_macros.h`)

**Comment by jneira** [comment]
> Check that stack based projects also work with the implicit configuration

For check that we could create a explicit `hie.yaml` with `cradle: {stack}`, right?

**Comment by jneira** [comment]
Building the branch is failing for me with stack due to `-Werror=unused-imports`:
```
haskell-ide-engine>     The import of `Haskell.Ide.Engine.Cradle' is redundant
haskell-ide-engine>       except perhaps to import instances from `Haskell.Ide.Engine.Cradle'
haskell-ide-engine>     To import instances alone, use: import Haskell.Ide.Engine.Cradle()
haskell-ide-engine>   |
haskell-ide-engine> 8 | import           Haskell.Ide.Engine.Cradle (findLocalCradle)
haskell-ide-engine>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
haskell-ide-engine>
haskell-ide-engine> D:\ws\haskell\haskell-ide-engine\app\MainHie.hs:19:1: error: [-Wunused-imports, -Werror=unused-imports]
haskell-ide-engine>     The import of `System.FilePath' is redundant
haskell-ide-engine>       except perhaps to import instances from `System.FilePath'
haskell-ide-engine>     To import instances alone, use: import System.FilePath()
haskell-ide-engine>    |
haskell-ide-engine> 19 | import           System.FilePath ((</>))
haskell-ide-engine>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Fixed with #53

**Comment by fendor** [comment]
> > Check that stack based projects also work with the implicit configuration
> 
> For check that we could create a explicit `hie.yaml` with `cradle: {stack}`, right?

I meant more manual testing. The ci is actively using only stack cradles, e.g. `cradle: { stack: {}} `. 
I am thinking of #32 for example.

**Comment by jneira** [comment]
> > Check that stack based projects also work with the implicit configuration
> 
> For check that we could create a explicit `hie.yaml` with `cradle: {stack}`, right?

I've tested in a simple project:
* With `cabal.project` and `stack.yaml` the cradle `Cabal-Helper-Cabal-V2` is succesfully loaded
* With `hie.yaml` =  `cradle: {stack}` the hie-bios `stack` cradle is succesfully loaded (with `cabal.project`present)
* With `stack.yaml` and no `cabal.project` the cradle `Cabal-Helper-Stack` is succesfully loaded

**Comment by jneira** [comment]
Not sure if it worths open an issue but hie doesnt create/use `C:\TEMP\hie.log` if the vscode log location option is empty. If i set the option it works (tested by @reactormonk and myself).

**Comment by lukel97** [comment]
@jneira for the vscode extension I changed it to go the output window by default when no log file is set, since I thought it was a bit confusing to have logs going to two separate places!

**Comment by jneira** [comment]
Oh, ok, agree it is better, but the output of the two previous sources are being merged in the same output window now?

**Comment by lukel97** [comment]
@jneira Yeah. Previously just the logm/debugm calls from hs-logger got sent to the log file. So now it’s all those calls, plus stderr and any child process stderr

**Comment by jneira** [comment]
I've run the test suite in my windows 10 with [`persist-virtual-file`](https://github.com/mpickering/haskell-ide-engine/pull/67) and [the fix for VFS keys on windows](https://github.com/alanz/haskell-lsp/pull/205): 
Results:
* Output: https://gist.github.com/jneira/6c00a2e78a1f3083b20070b4e77257f8
* Unit tests: two failing ones  
    * `HaRePlugin, hare plugin, Additional GHC API commands` 
 * Functional test: seven failing ones, but most of them due to slighty different output. Only two seems to be meaningful:
   * `HieBios.hie-bios reports errors in hie.yaml: uncaught exception: ConduitParserException`
   * `TypeDefinition, type definitions, find type-definition of type def in component`

Some of them are the same than the reported ones [here](https://github.com/mpickering/haskell-ide-engine/pull/67#issuecomment-558851974)

**Comment by jneira** [comment]
#71 would fix the `HieBios.hie-bios reports errors in hie.yaml: uncaught exception: ConduitParserException` failing test

**Comment by alanz** [comment]
I am beginning to wonder whether we should not remove HaRe from hie until after the hie-bios merge, and whatever we do with ghcide.

We do not currently have a good story for getting the project level module graph, and this means HaRe cannot work effectively. It has also been languishing for a long time waiting for underlying support, so I am not all that confident of its actual usefulness at the moment.

Thoughts?

**Comment by fendor** [comment]
Which features would be disabled by it? 
* rename
* find cross component, etc?

**Comment by alanz** [comment]
I am just starting to take a look at that, as a fork off mpickering/hie-bios

**Comment by mpickering** [comment]
I have now released `hie-bios-0.3`.

**Comment by fendor** [comment]
We made it into master!
